ZKUI 57-Add query params in the url

ChengYanJin opened the pull request:
Add the following props in the query params:
- showversions
- prefix
- versionId
- metadatasearch

With this PR, we centralize the  **listObject** action only in one place.

Note that this PR will change the behaviors:
- We will keep the current object selection when toggling the list versions.


Note that:
**We bumb the core-ui version with v0.28.0**

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/zenko-ui/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-57 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.1.2`


Please check the `Fix Version/s` of ZKUI-57, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

JBWatenbergScality commented on src/react/databrowser/objects/ObjectList.jsx:
```diff
@@ -114,6 +120,8 @@ export default function ObjectList({
+              dispatch(push(`${pathname}?${query.toString()}`));
               const type = isVersioningType
                 ? LIST_OBJECTS_S3_TYPE
                 : LIST_OBJECT_VERSIONS_S3_TYPE;
```
here and the line just below, do we still need to dispatch listObjects ? Can't we just rely on the fact that the url changing the Objects component will trigger the listing of the objects ?

JBWatenbergScality replied:
addressed by https://github.com/scality/zenko-ui/pull/215/commits/cf36e256ad62db4dfff8273954f3a2caadee60fb I guess

JBWatenbergScality commented on src/react/databrowser/objects/ObjectRow.jsx:
```diff
@@ -58,9 +61,17 @@ const Row = ({
+
   const handleClick = () => {
     dispatch(toggleAllObjects(false));
     dispatch(toggleObject(row.original.key, row.original.versionId));
```
I think this line may be dropped as the ObjectListTable will dispatch it when the url change

JBWatenbergScality commented on src/react/databrowser/objects/ObjectListTable.jsx:
```diff
@@ -78,6 +79,14 @@ export default function ObjectListTable({
+  const versionId = query.get('versionId');
+
+  useEffect(() => {
+    dispatch(toggleObject(objectKey, versionId ? versionId : undefined));
```
I see in the ObjectRow that before togglingAnObject it is dispatching `dispatch(toggleAllObjects(false));` do we nned to do so here as well ? What does `dispatch(toggleAllObjects(false));` exactly do ? I guess it is dis-selecting all the line of the table, if so I think we should likely not do that as it would loose the selection context of the user as soon as the user is checking one of the object _they_ selected

ChengYanJin replied:
The issue with removing the `dispatch(toggleAllObjects(false));` is that given an object is already selected, it would not be possible now to check another object, because as soon as I click on another object, its a multiple selection state :(

JBWatenbergScality commented on src/react/databrowser/objects/MetadataSearch.jsx:
```diff
@@ -69,13 +73,19 @@ const MetadataSearch = ({
       return;
     }
     dispatch(newSearchListing(bucketName, inputText));
```
I think this line may be dropped ?

ChengYanJin replied:
🤔  I don't think we should drop this line, it is the action to do the search objects.

JBWatenbergScality commented on src/react/databrowser/objects/MetadataSearch.jsx:
```diff
@@ -69,13 +73,19 @@ const MetadataSearch = ({
     e.stopPropagation();
     setHintsShown(false);
     setInputText('');
     dispatch(listObjects(bucketName, prefixWithSlash));
```
I think this line may be dropped ?

ChengYanJin replied:
🤔  I think if we really want to drop this line, then we probably need to add the value of metadatasearch query param in the dependency list where we dispatch the `listObjects` in the Objects.jsx

I am wondering, it is something we want?

```js
  useEffect(() => {
    dispatch(
      listObjects(
        bucketNameParam,
        prefixWithSlash,
        isShowVersions ? LIST_OBJECT_VERSIONS_S3_TYPE : LIST_OBJECTS_S3_TYPE,
      ),
    ).finally(() => setLoaded(true));
  }, [bucketNameParam, prefixWithSlash, dispatch, isShowVersions]);
  ```

ChengYanJin replied:
OK. So we took the decision to centralize the listObjects only in one place (Objects.jsx)

JBWatenbergScality commented on src/react/databrowser/objects/ObjectListTable.jsx:
```diff
@@ -83,11 +83,22 @@ export default function ObjectListTable({
+        versionId && isListVersion ? versionId : undefined,
+      ),
+    );
+    // Add `objectsLength` as dependency to make sure the toogleObject action is triggered after listObject
```
Then I guess we should wrap the above line with an `if (objectsLength) {}` so that it is more explicit and we don't need this comment to explain why we need that dependency

ChengYanJin replied:
Good point !!

JBWatenbergScality commented on src/react/databrowser/objects/ObjectListTable.jsx:
```diff
@@ -94,10 +94,13 @@ export default function ObjectListTable({
+    isListVersion,
+    objectsLength,
+    objectKey,
+    isListVersion ? versionId : null,
```
`isListVersion, versionId`

ChengYanJin replied:
OK. I think it would be easier to understand in this way.
Actually, we need to add a conditional dependency `versionId` in the dependency array.

The reason is that we are not listing versions, `versionId` should not appear in the dependency array. Otherwise when we patch the `versionId` to the URL query params it will trigger another **toggleObject** action with the versionId, where we don't have this object in the list..

```js
  const versionIdOrUndefined =
    isListVersion && versionId ? versionId : undefined;
  useEffect(() => {
    if (objectsLength > 0) {
      dispatch(
        toggleObject(
          isFolder ? `${objectKey}/` : objectKey,
          versionIdOrUndefined,
        ),
      );
    }
  }, [
    dispatch,
    isListVersion,
    objectsLength,
    objectKey,
    isFolder,
    versionIdOrUndefined,
  ]);
 ```

JBWatenbergScality replied:
Ok I see, I like this approach !

ChengYanJin committed:
ZKUI-57: Add metadatasearch in the query params

ChengYanJin committed:
ZKUI-57: Add usePrefixWithSlash hook to retrieve the prefixWithSlash

ChengYanJin committed:
ZKUI-57: Add react hook testing library

ChengYanJin committed:
ZKUI-57: Add prefix, showversions, versionId  in the query params

ChengYanJin committed:
ZKUI-57: Use usePrefixWithSlash hook to retrieve the prefix

ChengYanJin committed:
ZKUI-57: Patch versionId in query param in Properties

nicolas2bert commented:
Have not look at the code yet but I run the branch locally and found an issue with adding an object in a prefix (new folder). 
Also, it looks like we lose the context from the breadcrumb when moving into a prefix/folder.
When versioning is on, it looks like adding a new folder throws: `(Uncaught TypeError: Cannot read properties of undefined (reading 'length')`.
Also when select an object and click to Metadata or Tag tab on the right side, it looks like the query params get overwritten.

nicolas2bert commented:
Some of the issue I listed might not be related to your changes and we will need to create ticket for them if not.
(This one might be out of the scope) When I reload https://<UI>/buckets/bucket-test/objects?metadatasearch=key+like+%2Fjpg%24%2F), I was expecting to see `key like /jpg$/` value set in the metadata search input.
One idea could be that, before reviewing this code, we ask @Cuervino to review the UX to see if it aligns with the requirements. 
What do you think @ChengYanJin ?

ChengYanJin committed:
ZKUI-57: Fix the object tab selection

ChengYanJin committed:
ZKUI-57: Fix the breadcrumb followinng the URL changes

ChengYanJin committed:
ZKUI-57: Fix the error when listing versions

When we have only one object in the bucket, enable list version may not
keep the current object selection.

ChengYanJin committed:
ZKUI-57: don't patch the versionID when its not exist

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-57 contains:


* `1.1.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.1.3`


Please check the `Fix Version/s` of ZKUI-57, or the target
branch of this pull request.

ChengYanJin committed:
ZKUI-57: Fix the issue that we can't keep the current selected object

JBWatenbergScality commented on src/js/utils.js:
```diff
@@ -20,3 +21,11 @@ export function removeTrailingSlash(prefix: string): string {
+    result[key] = value;
+  }
+  return result;
+}
```
I think query can be typed using URLSearchParams. Also I think this is equivalent to `Object.fromEntries(query.entries())`

JBWatenbergScality commented on src/react/databrowser/objects/ObjectDetails.jsx:
```diff
@@ -67,21 +69,32 @@ function ObjectDetails({ toggled, listType }: Props) {
+  function removeFirstTabFromSearchQuery(queryObject: TabQueryObject) {
+    delete queryObject.tab;
+    return queryObject;
+  }
```
Calling this function would mutate the original object introducing a side effect which may be difficult to debug if later on we want to access queryObject.tab. I suggest to refactor it to a pure function by making a deep copy of queryObject or by trying to pass {...queryObject, tab: null} to the summary tab.

JBWatenbergScality commented on src/react/databrowser/objects/MetadataSearch.jsx:
```diff
@@ -58,11 +58,16 @@ const MetadataSearch = ({ isMetadataType, errorZenkoMsg }: Props) => {
 
+  useEffect(() => {
+    setInputText(searchInput);
+  }, []);
```
Instead of an effect I think this could be done when initiating input text state.
```js
const query = useQuery();
const searchInput = query.get('metadatasearch');
const [inputText, setInputText] = useState(searchInput || '');
```
That would avoid one useless rerender.

JBWatenbergScality reviewed (commented):
The build is failing due to 
`error  'useEffect' is defined but never used  no-unused-vars
 ` in src/react/databrowser/objects/MetadataSearch.jsx. Otherwise LGTM !

ChengYanJin committed:
ZKUI-57: Bump the core-ui version to 0.28.0

ChengYanJin committed:
ZKUI-57: Keep the metadatasearch result in the search input

ChengYanJin commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-57 contains:


* `1.1.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.1.3`


Please check the `Fix Version/s` of ZKUI-57, or the target
branch of this pull request.




*The following options are set:* **approve**

ChengYanJin commented:
Thanks for checking the behaviors on your side @nicolas2bert.  
I have checked with Valamir for this PR, and fixed some bugs.
Could you please take the time to review it again?

nicolas2bert reviewed (commented):
When I create a "folder" and enter it, the breadcrumb is not updated with the folder name.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **approve**

ChengYanJin commented:
> When I create a "folder" and enter it, the breadcrumb is not updated with the folder name.

Hi Nicolas, 
Good catch!! I fixed in f81f53d55fe8c14587d64f04e02453bc19a84f85

JBWatenbergScality commented on src/react/ui-elements/Breadcrumb.jsx:
```diff
@@ -28,18 +29,30 @@ const CustomBreadCrumb = styled(CoreUIBreadcrumb)`
     return [<label key="buckets">create bucket</label>];
   }
-  match = matchPath(pathname, { path: '/buckets/:bucketName/objects/*' });
+  match = matchPath(pathname, { path: '/buckets/:bucketName/objects*' });
```
I would introduce a new var named `matchObjectRoutes` instead of mutating the `match` one

gaudiauj commented on src/react/databrowser/objects/MetadataSearch.jsx:
```diff
@@ -68,14 +70,18 @@ const MetadataSearch = ({
     setHintsShown(false);
     setInputText('');
-    dispatch(listObjects(bucketName, prefixWithSlash));
+    // Remove the medatasearch from the query params
```
```suggestion
    // Remove the metadasearch from the query params
```

gaudiauj commented on src/react/databrowser/objects/ObjectDetails.jsx:
```diff
@@ -71,17 +71,25 @@ function ObjectDetails({ toggled, listType }: Props) {
+        <CustomTabs.Tab
+          label="Summary"
+          path={pathname}
+          query={{ ...queryObject, ...{ tab: null } }}
```
```suggestion
          query={{ ...queryObject, tab: null }}
```

gaudiauj commented on src/react/databrowser/objects/ObjectDetails.jsx:
```diff
@@ -71,17 +71,25 @@ function ObjectDetails({ toggled, listType }: Props) {
           label="Metadata"
           path={pathname}
-          query={{ tab: 'metadata' }}
+          query={{ ...queryObject, ...{ tab: 'metadata' } }}
```
```suggestion
          query={{ ...queryObject, tab: 'metadata'  }}
```

gaudiauj commented on src/react/databrowser/objects/ObjectDetails.jsx:
```diff
@@ -71,17 +71,25 @@ function ObjectDetails({ toggled, listType }: Props) {
+        <CustomTabs.Tab
+          label="Tags"
+          path={pathname}
+          query={{ ...queryObject, ...{ tab: 'tags' } }}
```
```suggestion
          query={{ ...queryObject, tab: 'tags' }}
```

gaudiauj commented on src/react/databrowser/objects/ObjectListTable.jsx:
```diff
@@ -73,11 +74,42 @@ export default function ObjectListTable({
+      query.set('prefix', key);
+      dispatch(push(`/buckets/${bucketName}/objects?${query.toString()}`));
     },
     [dispatch],
```
just for eslint to be happy, it shouldn't change the code beahavior
```suggestion
    [dispatch, query],
```

gaudiauj commented on src/react/databrowser/objects/ObjectRow.jsx:
```diff
@@ -58,14 +62,33 @@ const Row = ({
   return (
     <T.Row
-      isSelected={row.original.toggled}
+      isSelected={row.original.toggled || isRowSelected}
```
```suggestion
      isSelected={isRowSelected}
```

gaudiauj commented on src/react/databrowser/objects/ObjectRow.jsx:
```diff
@@ -58,14 +62,33 @@ const Row = ({
+    dispatch(push(`${pathname}?${query.toString()}`));
   };
-
+  const isRowSelected = isListVersions
+    ? row.original.versionId === versionId && objectKey === row.original.key
+    : objectKey === row.original.key;
```
```suggestion
  const isRowSelected = (isListVersions
    ? row.original.versionId === versionId && objectKey === row.original.key
    : objectKey === row.original.key) || row.original.toggled;
```

gaudiauj commented on src/react/databrowser/objects/Objects.jsx:
```diff
@@ -30,14 +33,40 @@ export default function Objects() {
+          prefixWithSlash,
+          isShowVersions ? LIST_OBJECT_VERSIONS_S3_TYPE : LIST_OBJECTS_S3_TYPE,
+        ),
+      ).finally(() => setLoaded(true));
```
i'm not sure the dispatch function work as a async function. Not sure that the finally will run after all actions are done.

ChengYanJin replied:
Since here I only have one action, it should be OK.

gaudiauj reviewed (approved):
just few comment about small code improvements. Otherwise look nice !

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.1`



The following branches will **NOT** be impacted:

* `development/1.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.1`



The following branches have **NOT** changed:

* `development/1.0`



Please check the status of the associated issue ZKUI-57.

Goodbye chengyanjin.
